### DISCUSSION TITLE
Is it possible to define Collection shapes?

### DISCUSSION THREAD

**Discussion by Grldk** [larastan/larastan#1659] [Q&A] [answered]
- larastan v2.6.2
- laravel v9.52.9
- phpstan v1.10.18
- phpstan level 6

Currently working to get our codebase error free on phpstan level 6, running into the following: I've got a Model class with a Collection `@property` for which I want to define a shape.

Lets say I have this code (simplified part of our actual code):

```php
/**
 * @property \Illuminate\Support\Collection|array{
 *  label: string,
 *  choices?: array{
 *      id: int,
 *      value: string,
 *  },
 * } $properties
 * /
class Field extends Model
{
    function test(): void
    {
        $label = $this->properties['label'];
    }
}
```
PHPStan says the type of `$label` is `mixed`, where I expect it to be `string`. But I cannot really find if this is possible, or if I should just use a plain array if I want to use shapes like this...

**Comment by mad-briller** [comment, marked as answer]
Yeah collection doesn't support shapes unfortunately

For this situation I typically use a class which extends Illuminate\Support\Fluent and has `@property` annotations

**Comment by szepeviktor** [comment]
Yes.

You could use a sovereign object instead of a non-list array.
Then you could say `$properties->getChoices()`!

[_source_](https://github.com/szepeviktor/very-clean-code#-arrays-versus-objects)

**Comment by canvural** [comment]
It *might* be possible to support it with an extension. Similar to https://github.com/nunomaduro/larastan/blob/73e5be5f5c732212ce6ca77ffd2753a136f36a23/src/ReturnTypes/ContainerArrayAccessDynamicMethodReturnTypeExtension.php But I'd not want this in core.
